### PR TITLE
TUN-1.4 Interface based IPv6 GRE Encapsulation

### PR DESCRIPTION
**Pull request by charantejag504**
TUN-1.4 Interface based IPv6 GRE Encapsulation

Validate the GRE configuration
Configure GRE Tunnel interfaces configuration options - Tunnel source, Tunnel Destination on Tunnel interface
Configure static route with IPv6 NH pointing to Tunnel interfaces
Validate the applied config did not report any errors
Enable the packet capture on ATE ingress port to verify the GRE header of uncapped traffic
Verify the tunnel interfaces counters to confirm the traffic encapsulation
After encapsulation, traffic should be load balanced/hash to all available L3 ECMP or LAG or combination of both features

This resolves https://github.com/openconfig/featureprofiles/issues/1601

### CODE REVIEW

File: proto/metadata.proto [hunk 1]
```diff
@@ -316,7 +316,11 @@ message Metadata {
     // Device does not support telemetry path
     // /qos/interfaces/interface/input/virtual-output-queues/voq-interface/queues/queue/state/dropped-pkts.
     bool qos_voq_drop_counter_unsupported = 103;
-
+    // Device does not support tunnel interfaces state paths
+    bool tunnel_state_path_unsupported = 104;
+    // Device does not support tunnel interfaces source and destination address config paths
+    bool tunnel_config_path_unsupported = 105;
```
Comment (sachendras) [inline proto/metadata.proto:322]: Please raise a deviation bug similar to b/299545397

File: proto/metadata.proto [hunk 2]
```diff
@@ -316,7 +316,11 @@ message Metadata {
     // Device does not support telemetry path
     // /qos/interfaces/interface/input/virtual-output-queues/voq-interface/queues/queue/state/dropped-pkts.
     bool qos_voq_drop_counter_unsupported = 103;
-
+    // Device does not support tunnel interfaces state paths
+    bool tunnel_state_path_unsupported = 104;
```
Comment (sachendras) [inline proto/metadata.proto:320]: Please raise a deviation bug similar to b/299545397

File: proto/metadata.proto (hunk 1 above)
Comment (charantejag504) [reply proto/metadata.proto]: updated bug

File: proto/metadata.proto (hunk 2 above)
Comment (charantejag504) [reply proto/metadata.proto]: updated bug

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/README.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+## Procedure
+Validate the GRE configuration
+GRE Tunnel interfaces configuration options
+Tunnel source
+Tunnel source should be able to configure with unnumbered interface address
+Tunnel Destination
+Directly on Tunnel interface or Tunnel Group
+Configure such 32 tunnel interfaces
+Filter match and action
```
Comment (sachendras) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/README.md:15]: This being an interface based Tunnel (and not Filter based tunnel), these Filters shouldnt exist?

Comment (charantejag504) [reply feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/README.md]: Filter related steps and paths are removed.

**Review by greg-dennis — CHANGES_REQUESTED** [review]
In addition, see all my comments on
https://github.com/openconfig/featureprofiles/pull/1999

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go [hunk 4]
```diff
@@ -0,0 +1,840 @@
...
+	otg "github.com/openconfig/ondatra/otg"
+	"github.com/openconfig/ygot/ygot"
+)
+
+func TestMain(m *testing.M) {
+	fptest.RunTests(m)
+}
+
+const (
+	Ipv6PrefixLen     = 24
```
[the remark this answers is not in the document]
Comment (charantejag504) [reply feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go]: Addressed

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go [hunk 5]
```diff
@@ -0,0 +1,840 @@
...
+	wantLoss := false
+	t.Run("Verify load balance and traffic drops with IPv6 flow via 8 tunnel", func(t *testing.T) {
+		t.Log("Verify load balance and traffic drops with IPv6 flow via 8 tunnel")
+		VerifyUnderlayOverlayLoadbalanceTest(t, p, dut1, rt, rt, d1p1, d1p2, d1p3, d1p4, rt1, rt2, rt2, rt3, 8, wantLoss)
+	})
+	captureTrafficStats(t, rt, otgConfig)
+
+}
+
+func ConfigureLoobackInterfaceWithIPv6address(t *testing.T, address string, dut *ondatra.DUTDevice) {
```
[the remark this answers is not in the document]
Comment (charantejag504) [reply feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go]: Addressed

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go [hunk 6]
```diff
@@ -0,0 +1,840 @@
...
+
```
[the remark this answers is not in the document]
Comment (charantejag504) [reply feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go]: Addressed

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go [hunk 7]
```diff
@@ -0,0 +1,840 @@
...
+			intfName: dp2.Name(),
+			ipAddr:   p.r0Intf2Ipv6Add,
+			Ipv6mask: p.ipv6Mask,
+		}, {
+			desc:     "R0_R1_1",
+			intfName: dp3.Name(),
+			ipAddr:   p.r0Intf3Ipv6Add,
+			Ipv6mask: p.ipv6Mask,
+		},
+		{
```
[the remark this answers is not in the document]
Comment (charantejag504) [reply feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go]: Addressed

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go [hunk 8]
```diff
@@ -0,0 +1,840 @@
...
+	max := expectedPerLinkPkts + (expectedPerLinkPkts * uint64(tolerance) / 100)
+
+	if min < stats && stats < max {
+		t.Logf("Traffic  %d is in expected range: %d - %d", stats, min, max)
+		t.Logf("Traffic Load balance Test Passed!")
+	} else {
+		t.Errorf("Traffic is expected in range %d - %d but got %d. Load balance Test Failed\n", min, max, stats)
+
+	}
+
```
[the remark this answers is not in the document]
Comment (charantejag504) [reply feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go:467]: Addressed

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go [hunk 9]
```diff
@@ -0,0 +1,840 @@
...
+func ConfigureLoobackInterfaceWithIPv6address(t *testing.T, address string, dut *ondatra.DUTDevice) {
+
+	// IPv6 address on lo0 interface
+	t.Logf("Push the Ipv6 address to lo0 interface :\n%s", dut.Vendor())
+	switch dut.Vendor() {
+	case ondatra.JUNIPER:
+		config := ConfigureAdditionalIPv6AddressonLoopback(address)
+		t.Logf("Push the CLI config:\n%s", config)
+
+		gnmiClient := dut.RawAPIs().GNMI().Default(t)
```
[the remark this answers is not in the document]
Comment (charantejag504) [reply feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go]: Addressed

**Comment by charantejag504** [comment]
> In addition, see all my comments on #1999

Addressed

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+
+		t.Logf("overlay static route via tunnel for an original IPv6 destination prefix")
+		for i, nextHop := range overlayIPv6Nh {
+			_, ipv6Destination := fetchNetworkAddress(t, encapInnerDesIpv6Network, interfacePlen6)
+			t.Logf("configuring static route in %s destination %s with next-hop %s", dut, ipv6Destination, nextHop)
+			configStaticRoute(t, dut, ipv6Destination, nextHop, strconv.Itoa(i))
+		}
+	})
+	t.Run("Configure OTG ports", func(t *testing.T) {
+		top := ate.OTG().NewConfig(t)
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go:154]: this function is deprecated. Just use gosnappi.NewConfig()

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go [hunk 11]
```diff
@@ -0,0 +1,467 @@
...
+}
+func sendTraffic(t *testing.T, ate *ondatra.ATEDevice) {
+	otg := ate.OTG()
+	t.Logf("Starting traffic")
+	otg.StartTraffic(t)
+	time.Sleep(time.Duration(trafficDuration) * time.Second)
+	t.Logf("Stop traffic")
+	otg.StopTraffic(t)
+}
+func configureOtgPorts(top gosnappi.Config, port *ondatra.Port, name string, mac string, ipv6Address string, ipv6Gateway string, ipv6Mask uint8) {
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go:252]: add blank line above, delete blank line below

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go [hunk 12]
```diff
@@ -0,0 +1,467 @@
...
+	networkWithMask := network.String() + "/" + strconv.Itoa(mask)
+	networkAlone := network.String()
+	//t.Logf("Network address : %s", networkWithMask)
+	return networkAlone, networkWithMask
+}
+func incrementAddress(t *testing.T, address string, i int, part string) string {
+	addr := net.ParseIP(address)
+	var oct int
+
+	if part == "network" {
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go:307]: use a switch statement. Can part be anything other than "network" or "host". If not, fail fatally in that case

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go [hunk 13]
```diff
@@ -0,0 +1,467 @@
...
+}
+
+func configStaticRoute(t *testing.T, dut *ondatra.DUTDevice, prefix string, nexthop string, index string) {
+	ni := oc.NetworkInstance{Name: ygot.String(deviations.DefaultNetworkInstance(dut))}
+	static := ni.GetOrCreateProtocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_STATIC, deviations.StaticProtocolName(dut))
+	sr := static.GetOrCreateStatic(prefix)
+	nh := sr.GetOrCreateNextHop(index)
+	nh.NextHop = oc.UnionString(nexthop)
+	gnmi.Update(t, dut, gnmi.OC().NetworkInstance(deviations.DefaultNetworkInstance(dut)).Protocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_STATIC, deviations.StaticProtocolName(dut)).Config(), static)
+
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go:401]: delete blank line

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go [hunk 14]
```diff
@@ -0,0 +1,467 @@
...
+			Val: &gpb.TypedValue{
+				Value: &gpb.TypedValue_AsciiVal{
+					AsciiVal: config,
+				},
+			},
+		}},
+	}
+	return gpbSetRequest
+}
+func configureNetworkInstance(t *testing.T, dut *ondatra.DUTDevice) {
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go:421]: add blank line above

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go [hunk 15]
```diff
@@ -0,0 +1,467 @@
...
+	t.Logf("Configure routing instance on dut")
+	dutConfPath := gnmi.OC().NetworkInstance(deviations.DefaultNetworkInstance(dut))
+	gnmi.Replace(t, dut, dutConfPath.Type().Config(), oc.NetworkInstanceTypes_NETWORK_INSTANCE_TYPE_DEFAULT_INSTANCE)
+}
+
+func verifyEcmpLoadBalance(t *testing.T, inital []uint64, final []uint64, flowCount int64, sharingIntfCont int64, firstintf int, wantLoss bool, lbTolerance int) {
+
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go:428]: delete blank line

**Review by greg-dennis — DISMISSED** [review]

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go (hunk 15 above)
Comment (charantejag504) [reply feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go]: deleted

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go (hunk 14 above)
Comment (charantejag504) [reply feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go:434]: added

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go (hunk 13 above)
Comment (charantejag504) [reply feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go]: deleted

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go (hunk 12 above)
Comment (charantejag504) [reply feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go]: moved to a switch statement

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go (hunk 11 above)
Comment (charantejag504) [reply feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go]: modifications made

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go (hunk 10 above)
Comment (charantejag504) [reply feature/experimental/tunnel/otg_tests/tunnel_interface_based_ipv6_gre_encapsulation_test/tunnel_interface_based_ipv6_gre_encapsulation_test.go]: moved to gosnappi.NewConfig()

**Comment by arulkumarsekar** [comment]
gcbrun

**Review by arulkumarsekar — APPROVED** [review]

**Pull request merged**